kube3d: increase discoverability

06kellyjac opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

I've searched for "k3d" previously using https://search.nixos.org but `kube3d` didn't show up so I've been maitaining my own copy of it for months :sweat_smile: 
Mentioning "k3d" in the `description` and `longDescription` should make it easier to find.

###### Things done

`nixpkgs-fmt`ed the file to fit with contributing guidelines.

Mentioned `k3d` in the `description` and `longDescription`

https://search.nixos.org can find things from either description but `nix search` just looks at the short description.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS (x86_64)
   - ~[ ] macOS~ Not on the platforms list
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

06kellyjac committed:
kube3d: nixpkgs-fmt the file to fit contributing guidelines

06kellyjac committed:
kube3d: add a longDescription to provide detail and help searches

search.nixos.org was able to show me `deno` when searching for
"executable" which was in the longDescription

06kellyjac committed:
kube3d: mention k3d in the short description

While I was able to find `deno` by searching for "executable" from it's
longDescription on search.nixos.org, I couldn't find it using `nix search`

`nix search` seems to only look at the short description so it's worth
including there too.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/346

06kellyjac committed:
kube3d: re-order the derivation and general cleanup

Moved k3sVersion to be a variable.
Converted buildFlagsArray to an array
Moved vendorSha256 closer to src
Moved doCheck between build and install related bits (like where the
phase happens)
Replaced stdenv.lib with lib

06kellyjac commented:
/marvin opt-in
/status needs_reviewer

06kellyjac commented:
Result of `nixpkgs-review pr 99563` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kube3d</li>
  </ul>
</details>

jlesquembre commented:
@06kellyjac thanks for the PR, I like your changes, but I have 2 questions:
- Why did you convert buildFlagsArray to an array?
- Why did you replaced stdenv.lib with lib?

About the first question, I found this related issue: https://github.com/NixOS/nixpkgs/issues/99200 but I still don't understand the reason. 

I'm not opposed to do that changes, I just want to understand the motivation behind them

06kellyjac commented:
> Why did you convert buildFlagsArray to an array?

IMO it makes more sense for me that something called an "Array" actually be an array. I'm not sure if there are any other advantages or disadvantages that are more substantial.
I've never had any issues like the one you mentioned. Just from scanning it I'm not sure what's wrong with the concat output

```bash
λ BOB=("a" "b" "c")
λ echo "${BOB[@]}"
a b c
```

```bash
λ BUILDFLAGSARRAY=("-tags" "netgo" "-ldflags='-w -X PACKAGE.gitTag=VERSION'")
λ echo "${BUILDFLAGSARRAY[@]}" # notwrapped in 'quotes'
-tags netgo -ldflags='-w -X PACKAGE.gitTag=VERSION'
```

`-tags netgo -ldflags='-w -X PACKAGE.gitTag=VERSION'` looks fine to me, and at the very least it's working for k3d

> Why did you replaced `stdenv.lib` with `lib`?

If `stdenv` is only being used for `stdenv.lib` then why not just grab `lib` directly like many other golang packages. If `stdenv` was being used for additional things it'd make sense to keep it like it was.

jlesquembre commented:
> IMO it makes more sense for me that something called an "Array" actually be an array. 

Totally agree with that, it's only that in the examples I remember, it was always a string. And if it ain't broke, don't fix it.
But not idea if there is a recommended way for that, I didn't find anything in the docs.

> If stdenv is only being used for stdenv.lib then why not just grab lib directly like many other golang packages.

It makes sense to me, I only wanted to know if there are some best practices about it.

To me, it's ok to merge your changes, I just wanted to know if there are some guide lines about that topics

06kellyjac commented:
Thanks for looking it all over

> I just wanted to know if there are some guide lines about that topics

No worries. I've not been using and packaging for nix too long so I'm also trying to work out some of the best practices :sweat_smile:

jlesquembre commented:
I tested the executable and all looks good.
/status needs_merger

SuperSandro2000 commented on pkgs/applications/networking/cluster/kube3d/default.nix:
```diff
@@ -1,42 +1,51 @@
+  meta = with lib; {
     homepage = "https://github.com/rancher/k3d";
-    description = "A helper to run k3s (Lightweight Kubernetes. 5 less than k8s) in a docker container";
+    description = "A helper to run k3s (Lightweight Kubernetes. 5 less than k8s) in a docker container - k3d";
```
```suggestion
    description = "A helper to run k3s (Lightweight Kubernetes. 5 less than k8s) in a docker container";
```

06kellyjac replied:
I added k3d to the short description because `nix search` uses it but doesn't look at `longDescription`

> `nix search` seems to only look at the short description so it's worth
including there too.

https://github.com/06kellyjac/nixpkgs/commit/83cd754e32c4651ce81887f7aca0336c53670407

kevincox replied:
I think this is fine with me. Sound good to leave it @SuperSandro2000?

SuperSandro2000 commented:
Can you please resolve the merge conflict?

06kellyjac commented:
Done
